File: samples/scala/zentasks/app/controllers/Tasks.scala
package controllers

import play.api._
import play.api.mvc._

import play.api.data.mapping._
import Rules._

import java.util.{Date}

import anorm._

import models._
import views._

/**
 * Manage tasks related operations.
 */
object Tasks extends Controller with Secured {

  /**
   * Display the tasks panel for this project.
   */
  def index(project: Long) = IsMemberOf(project) { _ => implicit request =>
    Project.findById(project).map { p =>
      val tasks = Task.findByProject(project)
      val team = Project.membersOf(project)
      Ok(html.tasks.index(p, tasks, team))
    }.getOrElse(NotFound)
  }

  val taskValidation = From[UrlFormEncoded] { __ =>
    import Rules._
    ((__ \ "title").read(notEmpty) ~
     (__ \ "dueDate").read(option(date("MM/dd/yy"))) ~
     (__ \ "assignedTo").read[Option[String]]).tupled
  }

  // -- Tasks

  /**
   * Create a task in this project.
   */
  def add(project: Long, folder: String) =  IsMemberOf(project) { _ => implicit request =>
    taskValidation
      .validate(request.body.asFormUrlEncoded.getOrElse(Map.empty))
      .fold(
        errors => BadRequest,
        {
          case (title, dueDate, assignedTo) =>
            val task =  Task.create(
              Task(NotAssigned, folder, project, title, false, dueDate, assignedTo)
            )
            Ok(html.tasks.item(task))
        }
      )
  }

  /**
   * Update a task
   */
  def update(task: Long) = IsOwnerOf(task) { _ => implicit request =>
    (Path \ "done").read(boolean)
      .validate(request.body.asFormUrlEncoded.getOrElse(Map.empty))
      .fold(
        errors => BadRequest,
        isDone => {
          Task.markAsDone(task, isDone)
          Ok
        }
      )
  }

  /**
   * Delete a task
   */
  def delete(task: Long) = IsOwnerOf(task) { _ => implicit request =>
    Task.delete(task)
    Ok
  }

  // -- Task folders

  /**
   * Add a new folder.
   */
  def addFolder = Action {
    Ok(html.tasks.folder("New folder"))
  }

  /**
   * Delete a full tasks folder.
   */
  def deleteFolder(project: Long, folder: String) = IsMemberOf(project) { _ => implicit request =>
    Task.deleteInFolder(project, folder)
    Ok
  }

  /**
   * Rename a tasks folder.
   */
  def renameFolder(project: Long, folder: String) = IsMemberOf(project) { _ => implicit request =>
    (Path \ "name").read(notEmpty)
      .validate(request.body.asFormUrlEncoded.getOrElse(Map.empty))
      .fold(
        errors => BadRequest,
        newName => {
          Task.renameFolder(project, folder, newName)
          Ok(newName)
        }
      )
  }

}

File: samples/java/zentasks/app/controllers/Tasks.java
package controllers;

import play.mvc.*;
import play.data.*;
import static play.data.Form.*;

import java.util.*;

import models.*;
import views.html.tasks.*;

/**
 * Manage tasks related operations.
 */
@Security.Authenticated(Secured.class)
public class Tasks extends Controller {

    /**
     * Display the tasks panel for this project.
     */
    public static Result index(Long project) {
        if(Secured.isMemberOf(project)) {
            return ok(
                index.render(
                    Project.find.byId(project),
                    Task.findByProject(project)
                )
            );
        } else {
            return forbidden();
        }
    }
  
    // -- Tasks
  
    /**
     * Create a task in this project.
     */ 
    public static Result add(Long project, String folder) {
        if(Secured.isMemberOf(project)) {
            Form<Task> taskForm = form(Task.class).bindFromRequest();
            if(taskForm.hasErrors()) {
                return badRequest();
            } else {
                return ok(
                    item.render(Task.create(taskForm.get(), project, folder), true)
                );
            }
        } else {
            return forbidden();
        }
    } 
  
    /**
     * Update a task
     */  
    public static Result update(Long task) {
        if(Secured.isOwnerOf(task)) {
            Task.markAsDone(
                task,
                Boolean.valueOf(
                    form().bindFromRequest().get("done")
                )
            );
            return ok();
        } else {
            return forbidden();
        }
    }
  
    /**
     * Delete a task
     */
    public static Result delete(Long task) {
        if(Secured.isOwnerOf(task)) {
            Task.find.ref(task).delete();
            return ok();
        } else {
            return forbidden();
        }
    }
  
    // -- Task folders

    /**
     * Add a new folder.
     */
    public static Result addFolder() {
        return ok(folder.render("New folder", new ArrayList<Task>()));
    }
  
    /**
     * Delete a full tasks folder.
     */
    public static Result deleteFolder(Long project, String folder) {
        if(Secured.isMemberOf(project)) {
            Task.deleteInFolder(project, folder);
            return ok();
        } else {
            return forbidden();
        }
    }
  
    /**
     * Rename a tasks folder.
     */
    public static Result renameFolder(Long project, String folder) {
        if(Secured.isMemberOf(project)) {
            return ok(
                Task.renameFolder(project, folder, form().bindFromRequest().get("name"))
            );
        } else {
            return forbidden();
        }
    }

}

